# Advanced Coding_Homework 1017

Look through Oakland 311 calls the same way we looked through Berkeley 311 calls. Note any interesting observations.  
Modify by **Jiyun Tsai** on Oct. 21, 2022  

In [3]:
import pandas as pd
import requests

## Download data

Download [311 call data](https://data.oaklandca.gov/Infrastructure/Service-requests-received-by-the-Oakland-Call-Cent/quth-gb8e) from the City of Oakland's Open Data Portal. 

In [4]:
url = 'https://data.oaklandca.gov/api/views/quth-gb8e/rows.csv?accessType=DOWNLOAD'
r = requests.get(url, allow_redirects=False) #api

# write the content of the request into a file called `oakland_311.csv`
open('oakland_311.csv', 'wb').write(r.content)

200449755

## Import data and Preview data structure

In [5]:
# Import the .csv data to panda dataframe, notice the data will update daily
oakland_311_original = pd.read_csv('oakland_311.csv')

# Preview data using head() and tail()
oakland_311_original.head(10)

REQUESTID            DATETIMEINIT          SOURCE  \
0    1177440  01/11/2022 09:54:54 AM  Phone or Email   
1    1236100  09/04/2022 11:14:03 PM     SeeClickFix   
2    1236280  09/06/2022 09:52:30 AM           Phone   
3    1237256  09/09/2022 12:18:50 PM     SeeClickFix   
4    1238127  09/12/2022 11:02:06 PM     SeeClickFix   
5    1240181  09/20/2022 11:02:32 AM           Phone   
6    1241342  09/24/2022 09:26:19 AM     SeeClickFix   
7    1241521  09/25/2022 03:43:24 PM     SeeClickFix   
8    1241594  09/26/2022 08:43:26 AM           Phone   
9    1004444  04/24/2020 10:51:22 AM           Phone   

                                  DESCRIPTION   REQCATEGORY  \
0                                 Tree - Down         TREES   
1                         Hazardous Materials          FIRE   
2                           Sidewalk - Damage  ROW_STREETSW   
3                              Tree - Pruning         TREES   
4                       Parking - Enforcement       PARKING   
5                             Street Cleaning          KOCB   
6                             Litter in Parks         PARKS   
7  Illegal Dumping - debris, appliances, etc.       ILLDUMP   
8            Park - Tot Lots, Tables, Benches     BLDGMAINT   
9                         Homeless Encampment      HE_CLEAN   

                                  REQADDRESS                         STATUS  \
0                                        NaN                       WOCREATE   
1   (37.79110003129922, -122.22179990095928)                           OPEN   
2    (37.77609653732435, -122.1972178270711)                           OPEN   
3   (37.77617202897651, -122.18976602952326)                       UNFUNDED   
4  (37.764640839375126, -122.20173973510026)  EVALUATED - NO FURTHER ACTION   
5   (37.76861412209682, -122.18015232119743)                           OPEN   
6    (37.78529824288906, -122.1964578593114)                         CLOSED   
7   (37.78225856800642, -122.24048462722071)                         CANCEL   
8   (37.83277651750851, -122.27612014290075)                         CLOSED   
9  (37.794394119983984, -122.25283604613712)                         CANCEL   

  REFERREDTO          DATETIMECLOSED          SRX          SRY  \
0        NaN                     NaN          NaN          NaN   
1        NaN                     NaN  6064165.540  2115096.007   
2        NaN                     NaN  6071168.490  2109504.060   
3        NaN                     NaN  6073322.331  2109492.573   
4        NaN                     NaN  6069785.951  2105357.222   
5        NaN                     NaN  6076051.000  2106691.000   
6        NaN  09/29/2022 09:07:38 AM  6071448.831  2112850.067   
7        NaN                     NaN  6058707.478  2111977.060   
8        NaN  09/30/2022 02:57:48 PM  6048759.962  2130561.993   
9        NaN                     NaN  6055221.383  2116461.724   

  COUNCILDISTRICT BEAT            PROBADDRESS     City State  
0             NaN  02Y            McCRAE POND  Oakland    CA  
1            CCD5  21X   E 27TH ST & BARRY PL  Oakland    CA  
2            CCD6  28X           2600 COLE ST  Oakland    CA  
3            CCD6  29X          2901 56TH AVE  Oakland    CA  
4            CCD6  26Y          1291 58TH AVE  Oakland    CA  
5            CCD6  30Y  68TH AV & BANCROFT AV  Oakland    CA  
6            CCD6  28X        COURTLAND CREEK  Oakland    CA  
7            CCD5  19X               22ND AVE  Oakland    CA  
8            CCD1  10Y            LINDEN PARK  Oakland    CA  
9            CCD2  19X           CLINTON PARK  Oakland    CA

In [6]:
# tail() to look at the last rows of the data
oakland_311_original.tail(10) ## there are 935733 rows as of Oct.21, 2022

REQUESTID            DATETIMEINIT       SOURCE  \
947465     993167  03/05/2020 01:08:36 PM  SeeClickFix   
947466     999645  04/03/2020 07:34:54 AM  SeeClickFix   
947467     994966  03/11/2020 03:09:02 PM  SeeClickFix   
947468     996031  03/16/2020 03:28:37 PM    Voicemail   
947469     999454  04/02/2020 09:08:26 AM  SeeClickFix   
947470     992269  03/03/2020 08:13:59 AM        Email   
947471     998818  03/30/2020 11:30:19 AM  SeeClickFix   
947472     998739  03/30/2020 08:56:33 AM  SeeClickFix   
947473     994068  03/09/2020 09:47:47 AM  SeeClickFix   
947474     997364  03/23/2020 10:31:25 AM        Phone   

                                              DESCRIPTION    REQCATEGORY  \
947465  Graffiti on Street, Street Light, Traffic Signal,       GRAFFITI   
947466                        Parking - Abandoned Vehicle         POLICE   
947467                TE - Traffic Safety (non-emergency)  TRAFFIC_ENGIN   
947468                              Trash - Missed Pickup      RECYCLING   
947469         Illegal Dumping - debris, appliances, etc.        ILLDUMP   
947470         Illegal Dumping - debris, appliances, etc.        ILLDUMP   
947471  Graffiti on a City Building (Library, Rec Center)      BLDGMAINT   
947472  Litter - Street Litter Container - Overflowing...        ILLDUMP   
947473         Illegal Dumping - debris, appliances, etc.        ILLDUMP   
947474         Illegal Dumping - debris, appliances, etc.        ILLDUMP   

                                       REQADDRESS  \
947465  (37.812651478894125, -122.29233517644725)   
947466  (37.820069233779755, -122.26071665430794)   
947467   (37.78464997936016, -122.18749631448534)   
947468   (37.82307316593434, -122.27606571784146)   
947469   (37.84259189198031, -122.25164877817319)   
947470  (37.768215040818085, -122.19858681924846)   
947471   (37.81143000137956, -122.26130549877647)   
947472   (37.82441723346091, -122.25230276622392)   
947473  (37.773025041459675, -122.18271682376431)   
947474   (37.74945634259511, -122.18104057382614)   

                               STATUS REFERREDTO          DATETIMECLOSED  \
947465                         CLOSED        NaN  03/09/2020 03:16:57 PM   
947466                         CLOSED        NaN  05/14/2020 09:45:07 AM   
947467  EVALUATED - NO FURTHER ACTION        NaN                     NaN   
947468                         CLOSED        NaN  03/18/2020 01:01:25 PM   
947469                         CLOSED        NaN  04/03/2020 04:59:23 PM   
947470                         CLOSED        NaN  03/04/2020 10:08:26 PM   
947471                         CLOSED        NaN  04/10/2020 11:10:45 AM   
947472                         CANCEL        NaN                     NaN   
947473                         CLOSED        NaN  03/10/2020 08:19:03 PM   
947474                         CLOSED        NaN  03/24/2020 08:10:16 PM   

                SRX          SRY COUNCILDISTRICT BEAT  \
947465  6043937.629  2123324.593            CCD3  05Y   
947466  6053120.595  2125851.721            CCD3  08X   
947467  6074033.871  2112567.240            CCD4  25X   
947468  6048708.627  2127029.100            CCD3  06X   
947469  6055892.846  2134002.200            CCD1  12Y   
947470  6070720.808  2106641.892            CCD6  27Y   
947471  6052891.339  2122709.727            CCD3  14X   
947472  6055580.128  2127389.044            CCD1  09X   
947473  6075338.737  2108310.166            CCD6  29X   
947474  6075668.999  2099721.001            CCD7  33X   

                                          PROBADDRESS     City State  
947465                                   1500 16TH ST  Oakland    CA  
947466                                  3054 BROOK ST  Oakland    CA  
947467                                  3727 BUELL ST  Oakland    CA  
947468                                 3233 Market St  Oakland    CA  
947469                                5505 COLLEGE AV  Oakland    CA  
947470                         57TH AVE & SCOVILLE ST  Oa

In [7]:
# Use info() to see types/structures of the dataframe

oakland_311_original.info() ## there are 16 columns, if the non-null count is lesser than 935733(as of Oct.21, 2022)
                            ## it means there are some missing values in the columns (id matches the total which is good)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 947475 entries, 0 to 947474
Data columns (total 16 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   REQUESTID        947475 non-null  int64  
 1   DATETIMEINIT     947475 non-null  object 
 2   SOURCE           947475 non-null  object 
 3   DESCRIPTION      947475 non-null  object 
 4   REQCATEGORY      947386 non-null  object 
 5   REQADDRESS       929892 non-null  object 
 6   STATUS           947473 non-null  object 
 7   REFERREDTO       84938 non-null   object 
 8   DATETIMECLOSED   683200 non-null  object 
 9   SRX              929892 non-null  float64
 10  SRY              929892 non-null  float64
 11  COUNCILDISTRICT  902508 non-null  object 
 12  BEAT             905241 non-null  object 
 13  PROBADDRESS      947023 non-null  object 
 14  City             947475 non-null  object 
 15  State            947475 non-null  object 
dtypes: float64(2), int64(1), object(13)
me

## Clean data

In [8]:
# Copy the dataframe so that we don't mess up with the original data
oakland_311 = oakland_311_original.copy()

In [9]:
oakland_311['REQUESTID'].nunique() ## see how many unique cases are there

947475

In [10]:
len(oakland_311) ## the unique number matches the len that means each row has unique id

947475

In [11]:
# Ways to check duplicated IDs
oakland_311[oakland_311['REQUESTID'].duplicated()] ## check if there are duplicated REQUESTID just in case

Empty DataFrame
Columns: [REQUESTID, DATETIMEINIT, SOURCE, DESCRIPTION, REQCATEGORY, REQADDRESS, STATUS, REFERREDTO, DATETIMECLOSED, SRX, SRY, COUNCILDISTRICT, BEAT, PROBADDRESS, City, State]
Index: []

In [12]:
assert len(oakland_311) == oakland_311['REQUESTID'].nunique() ## nothing happened, True

### Convert date type and calculate execute time
Use `df['column_name'] = pd.to_datetime(df['column_name'])`

In [13]:
# Convert 'DATETIMEINIT' and 'DATETIMECLOSED' to correct date type
oakland_311['DATETIMEINIT'].min()
oakland_311['DATETIMEINIT'].max() ## examine the min() and max() of the date and they are wrong

'12/31/2021 12:57:32 AM'

In [14]:
oakland_311['DATETIMEINIT'] = pd.to_datetime(oakland_311['DATETIMEINIT'])

In [15]:
oakland_311['DATETIMECLOSED'] = pd.to_datetime(oakland_311['DATETIMECLOSED'])

In [16]:
oakland_311.info() ## the date type has changed

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 947475 entries, 0 to 947474
Data columns (total 16 columns):
 #   Column           Non-Null Count   Dtype         
---  ------           --------------   -----         
 0   REQUESTID        947475 non-null  int64         
 1   DATETIMEINIT     947475 non-null  datetime64[ns]
 2   SOURCE           947475 non-null  object        
 3   DESCRIPTION      947475 non-null  object        
 4   REQCATEGORY      947386 non-null  object        
 5   REQADDRESS       929892 non-null  object        
 6   STATUS           947473 non-null  object        
 7   REFERREDTO       84938 non-null   object        
 8   DATETIMECLOSED   683200 non-null  datetime64[ns]
 9   SRX              929892 non-null  float64       
 10  SRY              929892 non-null  float64       
 11  COUNCILDISTRICT  902508 non-null  object        
 12  BEAT             905241 non-null  object        
 13  PROBADDRESS      947023 non-null  object        
 14  City             947

In [17]:
# Calculate 311 execute time by 'DATETIMECLOSED' - 'DATETIMEINIT'
oakland_311['Execute_Time'] = oakland_311['DATETIMECLOSED'] - oakland_311['DATETIMEINIT']

In [18]:
oakland_311['Execute_Time'].max() # whao, that's more than 10 years...

Timedelta('4293 days 21:59:22')

In [19]:
# Check columns if there are unusual data(but also see summary of each column):

# DATETIMEINIT: have changed the date type
# SOURCE: check if there's non-categorized description
# REQCATEGORY: check if there's non-categorized description
# STATUS: check if there's non-categorized description
# DATETIMECLOSED: have changed the date type
# COUNCILDISTRICT: from 2009 to 2022, there are seven districts in Oakland, check if there's number outside of 1 to 7 (and NAs)

### SOURCE
SOURCE = Indicates if the request was received via SeeClickFix or the OAK 311 mobile app, via the legacy Report A Problem website, or via (Phone or Email).

In [20]:
oakland_311['SOURCE'].describe() ## 'unique == 8' shows that there are 8 different values in this column

count          947475
unique              8
top       SeeClickFix
freq           449294
Name: SOURCE, dtype: object

In [21]:
# Also tried oakland_311.groupby("SOURCE").count() but this looks too messy
## SeeClickFix, Phone, Phone or Email, Website, Email, Voicemail, Other, and City Attorney
## SeeClickFix has the most count: 441478
## City Attorney has the least count: 196

oakland_311['SOURCE'].value_counts()

SeeClickFix       449294
Phone             192106
Phone or Email    152413
Website            77278
Email              58085
Voicemail          12926
Other               5177
City Attorney        196
Name: SOURCE, dtype: int64

### REQCATEGORY
REQCATEGORY = Request Category

In [22]:
## 'count' == 935645 shows that there are some missing values that we can take a closer look
## 'unique' == 35 means that there are a total 35 Request Category
## 'ILLDUMP' has the most frequency of 336320 (maybe stands for illegal dump?)

oakland_311['REQCATEGORY'].describe()

count      947386
unique         35
top       ILLDUMP
freq       339647
Name: REQCATEGORY, dtype: object

In [23]:
## Note that 'POLICE' has the third highest count, a total of 75566 (something to look closer later)

oakland_311['REQCATEGORY'].value_counts() ## to see detail counts of every category in this column

ILLDUMP           339647
OTHER              84046
BLDGMAINT          75690
POLICE             75561
STREETSW           52111
GRAFFITI           39555
ELECTRICAL         36319
RECYCLING          31714
TREES              31460
TRAFFIC            24466
PARKS              23917
TRAFFIC_ENGIN      16955
SEWERS             16027
PARKING            15672
DRAINAGE           13687
HE_CLEAN           11377
METER_REPAIR       10500
ROW_INSPECTORS      9939
KOCB                8288
ROW_STREETSW        7336
FIRE                6161
CUT_CLEAN           4801
ROW                 4371
ABANDONED AUTO      3970
HOMELESS EMT        1486
ENVIRON_ENF          610
WATERSHED            591
CW_DIT_GIS           562
FACILITIES           329
SIDESHOWS            156
VEGCONTR              39
SURVEY                36
GIS                    4
LAB                    2
OPD                    1
Name: REQCATEGORY, dtype: int64

### STATUS
The status as of date of upload. *PENDING:* Request received for review *OPEN:* Request assigned to work unit *WOCREATE:* Work Order created for work to be done (not all work units use this status, so Open service requests may also have work orders created for work to be done) *CLOSED:* Request is resolved *REFERRED:* Oakland Call Center received the request and forwarded it to another entity (see REFERREDTO) *UNFUNDED:* Request is for an service that the City is not funded to provide *CANCEL:* Request was canceled, generally because it was created erroneously, it was used for testing/training, or it is a request that was received that duplicated an existing unresolved request *WAITING ON CUSTOMER:* Request involves a process where customer needs to provide information requested by staff *EVALUATED - NO FURTHER ACTION:* Staff has evaluated the service request and no further action will be taken (e.g., limited resources, temporary issue, staff alerted to ongoing issue, professional judgement says to leave situation as-is, unable to provide requested action) *GONE ON ARRIVAL:* Staff went to location and was unable to verify the issue

In [24]:
# Notice there's a 'CANCEL' (count == 122473) and 'Cancel' (count == 2)

oakland_311['STATUS'].value_counts()

CLOSED                           680809
CANCEL                           123803
REFERRED                          69765
OPEN                              42091
UNFUNDED                          10731
EVALUATED - NO FURTHER ACTION      6600
GONE ON ARRIVAL                    5251
WOCREATE                           4454
PENDING                            3703
WAITING ON CUSTOMER                 264
Cancel                                2
Name: STATUS, dtype: int64

In [25]:
# Identify those two rows with 'STATUS' writing 'Cancel' to maybe change them to 'CANCEL' or call to confirm
## Use df.loc[df['col1'] == value] syntax

oakland_311.loc[oakland_311['STATUS'] == 'Cancel'] ## REQUESTID '734223' and '779834' have Cancel instead of CANCEL

REQUESTID        DATETIMEINIT       SOURCE  \
675524     734223 2017-05-24 08:58:59  SeeClickFix   
746734     779834 2017-12-13 19:33:05      Website   

                          DESCRIPTION REQCATEGORY  \
675524               Code Enforcement       OTHER   
746734  Streets - Potholes/Depression    STREETSW   

                                      REQADDRESS  STATUS         REFERREDTO  \
675524   (37.80848726830187, -122.2562403221613)  Cancel  CODE ENFORCEMENT,   
746734  (30.009899650831787, -141.2187509893542)  Cancel                NaN   

       DATETIMECLOSED        SRX        SRY COUNCILDISTRICT BEAT  \
675524            NaT  6054334.2  2121610.9            CCD3  14X   
746734            NaT        0.0        0.0             NaN  NaN   

                               PROBADDRESS     City State Execute_Time  
675524                  458-468 PERKINS ST  Oakland    CA          NaT  
746734  WILLOW BETWEEN 24TH ST AND 26TH ST  Oakland    CA          NaT

In [26]:
# Modify 'Cancel' to 'CANCEL'
## df['column'] = df['column'].str.replace('original value', 'new value')

oakland_311['STATUS'] = oakland_311['STATUS'].str.replace('Cancel', 'CANCEL')

In [27]:
# Check if changed sucessfully

oakland_311.loc[oakland_311['STATUS'] == 'Cancel'] ## there's no more 'Cancel' values in this column
# OR
oakland_311['STATUS'].value_counts() ## the original 2 rows of 'Cancel' have been added to 'CANCEL'

CLOSED                           680809
CANCEL                           123805
REFERRED                          69765
OPEN                              42091
UNFUNDED                          10731
EVALUATED - NO FURTHER ACTION      6600
GONE ON ARRIVAL                    5251
WOCREATE                           4454
PENDING                            3703
WAITING ON CUSTOMER                 264
Name: STATUS, dtype: int64

### COUNCILDISTRICT
COUNCILDISTRICT = City Council District of the REQADDRESS

In [28]:
# Council district should be numbers from 1 to 7
## Okay... it's a mess
## It shows there are 20 unique values in this column
## The total count is 891105 (there are many missing values) that possibly associated with address

oakland_311['COUNCILDISTRICT'].describe()

count     902508
unique        14
top         CCD3
freq      219119
Name: COUNCILDISTRICT, dtype: object

In [29]:
## Except the expected '1 to 7' values, there are 'CCD1 to CCD7' and some lowercase ccds, typo CDD, and 'District 4'
## Need to clean up but also need to know what CCD means

oakland_311['COUNCILDISTRICT'].value_counts()

CCD3                 219119
CCD7                 135208
CCD2                 132139
CCD1                 109799
CCD5                 106932
CCD6                 100847
CCD4                  98239
Piedmont                217
ccd4                      3
CCD21650 18TH AVE         1
CDD2                      1
District 4                1
ccd1                      1
ccd7                      1
Name: COUNCILDISTRICT, dtype: int64

## More descriptive statistics

In [30]:
# Execute_Time
## Note that there might be wrong values since the maxium of Execute_Time seems weird

In [31]:
# Mean
## Note that mean will get affected by outliers (if there are any)

oakland_311['Execute_Time'].mean()

Timedelta('51 days 23:19:47.531778397')

In [32]:
# Median
## The difference between mean and median gives an idea if we plot the Execute_Time out, what the distribution might look like

oakland_311['Execute_Time'].median()

Timedelta('4 days 22:07:44')

In [33]:
# Sort Execute_Time in descending way can show if there are outliers
## Okay... it seems that there are many cases that took them more than 10 years to finish...

oakland_311.sort_values(by=['Execute_Time'], ascending=False).head(15)

REQUESTID        DATETIMEINIT          SOURCE  \
285235     323595 2009-12-09 14:30:28  Phone or Email   
298351     356737 2011-01-07 09:10:01  Phone or Email   
302235     357071 2011-01-10 16:26:10  Phone or Email   
310100     357080 2011-01-11 07:44:49  Phone or Email   
306199     359977 2011-02-05 13:58:31         Website   
313834     360808 2011-02-14 09:28:36  Phone or Email   
297846     360135 2011-02-07 13:05:06         Website   
321963     368213 2011-04-27 14:42:37  Phone or Email   
327186     366116 2011-04-08 09:33:45  Phone or Email   
302809     358429 2011-01-24 12:39:20  Phone or Email   
11565      384229 2011-10-04 12:30:38         Website   
327931     368147 2011-04-27 10:39:37  Phone or Email   
327396     378532 2011-08-10 09:32:02  Phone or Email   
319001     373262 2011-06-16 15:20:54  Phone or Email   
318299     376435 2011-07-20 16:05:23  Phone or Email   

                          DESCRIPTION   REQCATEGORY  \
285235    Street Medians (Landscaped)         PARKS   
298351              Sidewalk - Damage  ROW_STREETSW   
302235              Sidewalk - Damage  ROW_STREETSW   
310100              Sidewalk - Damage  ROW_STREETSW   
306199              Sidewalk - Damage  ROW_STREETSW   
313834              Sidewalk - Damage  ROW_STREETSW   
297846              Sidewalk - Damage  ROW_STREETSW   
321963              Sidewalk - Damage  ROW_STREETSW   
327186              Sidewalk - Damage  ROW_STREETSW   
302809              Sidewalk - Damage  ROW_STREETSW   
11565   Streets - Potholes/Depression      STREETSW   
327931              Sidewalk - Damage  ROW_STREETSW   
327396              Sidewalk - Damage  ROW_STREETSW   
319001              Sidewalk - Damage  ROW_STREETSW   
318299              Sidewalk - Damage  ROW_STREETSW   

                                       REQADDRESS  STATUS REFERREDTO  \
285235  (37.822063947421924, -122.20671960118034)  CLOSED        NaN   
298351   (37.83381506361111, -122.22294168117693)  CLOSED        NaN   
302235   (37.77400051063402, -122.21222817106961)  CLOSED        NaN   
310100   (37.80453678977035, -122.23308765745459)  CLOSED        NaN   
306199   (37.83247777213932, -122.26208613791599)  CLOSED        NaN   
313834  (37.819594099154294, -122.25024335024222)  CLOSED        NaN   
297846  (37.843924050913586, -122.25830364202436)  CLOSED        NaN   
321963   (37.832717255708985, -122.2600975536424)  CLOSED        NaN   
327186   (37.83233483511009, -122.26216296319723)  CLOSED        NaN   
302809  (37.798803786784994, -122.18900833833311)  CLOSED        NaN   
11565    (37.78765877539137, -122.20094834312434)  CLOSED        NaN   
327931   (37.84820595299595, -122.25846970878077)  CLOSED        NaN   
327396   (37.76311547167693, -122.18858097466278)  CLOSED        NaN   
319001   (37.84468789233125, -122.26431985105448)  CLOSED        NaN   
318299  (37.762772892453214, -122.18776010072135)  CLOSED        NaN   

            DATETIMECLOSED          SRX          SRY COUNCILDISTRICT BEAT  \
285235 2021-09-11 12:29:50  6068728.000  2126289.000            CCD4  22X   
298351 2022-10-07 10:34:22  6064122.000  2130653.000            CCD1  13X   
302235 2022-10-07 10:58:46  6066817.000  2108820.000            CCD5  27X   
310100 2022-10-07 12:08:53  6060995.000  2120048.000            CCD2  15X   
306199 2022-08-09 09:26:19  6052810.163  2130376.620            CCD1  12X   
313834 2022-08-16 12:29:16  6056142.000  2125622.000            CCD2  14Y   
297846 2022-06-07 13:43:22  6053980.649  2134523.222            CCD1  12Y   
321963 2022-07-20 15:13:58  6053386.000  2130453.000            CCD1  12X   
327186 2022-06-20 10:17:57  6052787.000  2130325.000            CCD1  12X   
302809 2022-03-24 10:50:54  6073690.000  2117728.000            CCD4  25X   
11565  2022-11-04 16:01:49  6070167.000  2113733.000            CCD4  24Y   
327931 2022-05-20 14:16:43  6053962.000  2136083.000            CCD1  12Y   
327396 2022-08-22 09:50:43  6073579.000  2104733.000        

## What seems wrong

- DATETIMEINIT and DATETIMECLOSED columns have incorrect date format (fixed)
- Execute_Time columm has a maxium value of Timedelta('4293 days 21:59:22') which seems off or could be the outlier
- STATUS column has wrong value 'Cancel' that should be changed to 'CANCEL' (fixed)
- COUNCILDISTRICT column is a mess that need further clarification with the Oakland City as described above

## Export clean data

In [35]:
oakland_311.to_csv('oakland_311_clean.csv', index=False)